be/src/olap/rowset/segment_v2/inverted_index/query/conjunction_query.cpp

```diff
@@ -0,0 +1,162 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "conjunction_query.h"
+
+namespace doris {
+
+ConjunctionQuery::ConjunctionQuery(IndexReader* reader)
+        : _reader(reader), _indexVersion(reader->getIndexVersion()) {}
+
+ConjunctionQuery::~ConjunctionQuery() {
+    for (auto& term : _terms) {
+        if (term) {
+            _CLDELETE(term);
+        }
+    }
+    for (auto& termDoc : _termDocs) {
+        if (termDoc) {
+            _CLDELETE(termDoc);
+        }
+    }
+}
+
+void ConjunctionQuery::add(const std::wstring& fieldName, const std::vector<std::wstring>& wterms) {
+    if (wterms.size() < 1) {
+        _CLTHROWA(CL_ERR_IllegalArgument, "ConjunctionQuery::add: terms.size() < 1");
+    }
+
+    std::vector<TermIterator> iterators;
+    for (auto& wterm : wterms) {
+        Term* t = _CLNEW Term(fieldName.c_str(), wterm.c_str());
+        _terms.push_back(t);
+        TermDocs* termDoc = _reader->termDocs(t);
+        _termDocs.push_back(termDoc);
+        iterators.emplace_back(termDoc);
+    }
+
+    std::sort(iterators.begin(), iterators.end(), [](const TermIterator& a, const TermIterator& b) {
+        return a.docFreq() < b.docFreq();
+    });
+
+    if (iterators.size() == 1) {
+        _lead1 = iterators[0];
+    } else {
+        _lead1 = iterators[0];
+        _lead2 = iterators[1];
+        for (int32_t i = 2; i < _terms.size(); i++) {
+            _others.push_back(iterators[i]);
+        }
+    }
+
+    if (iterators.size() >= 2) {
+        int32_t little = iterators[0].docFreq();
+        int32_t big = iterators[iterators.size() - 1].docFreq();
+        if (little == 0) {
+            _useSkip = true;
+        } else if ((big / little) > _conjunction_ratio) {
+            _useSkip = true;
+        }
+    }
+}
+
+void ConjunctionQuery::search(roaring::Roaring& roaring) {
+    if (_lead1.isEmpty()) return;
+
+    if (_indexVersion == IndexVersion::kV0) {
```

Comment:
can be merged to set _use_skiplist and just select search_by_bitmap/search_by_skiplist by _use_skiplist